src/openfermion/measurements/_probability_distributions.py

```diff
@@ -0,0 +1,379 @@
+    """
+    Function to make the matrices that define multiplication
+    by sin(theta) and cos(theta). We store these as sparse matrices
+    in CSR format, and initialize them in (row,col,data) format.
+    We assume here that self.max_n is less than self.num_freqs.
+    """
+
+    # matrices[n][a] contains matrix M^a(n+1)
+    # following arXiv:1809.09697, app.C
+    matrices = []
+
+    for index in range(1, max_n+1):
+
+        # We generate the matrices ourselves in
+        # COO format, and then create a
+        # CSR matrix directly
+
+        cos_matrix, sin_matrix = make_fourier_matrix(index, num_freqs)
+
+        # Append to list
+        matrices.append([cos_matrix, sin_matrix])
+    return matrices
+
+def make_fourier_matrix(nrot, num_freqs):
+    """
+    Makes matrices that sends a Fourier representation of p(phi)
+    to the Fourier representation of cos^2(n*phi/2+beta/2)*p(phi).
+
+    See App. C in arXiv:1809.09697
+
+    Args:
+        nrot (n): the number of rotations n in the above formula.
+    Returns:
+        cos_matrix (scipy.sparse.csr_matrix) matrix for multiplication
+            by cos(phi)
+        sin_matrix (scipy.sparse.csr_matrix) matrix for multiplication
+            by sin(phi)
+    """
+
+    # Initialize with terms for when j = 0
+    data_cos = [2]
+    row_ind_cos = [2*nrot]
+    col_ind_cos = [0]
+
+    data_sin = [-2]
+    row_ind_sin = [2*nrot-1]
+    col_ind_sin = [0]
+
+    # First do terms when j < n
+    for j in range(1, nrot):
+
+        data_cos += [1, 1, -1, 1]
+        row_ind_cos += [2*j+2*nrot-1, 2*j+2*nrot, 2*nrot-2*j-1, 2*nrot-2*j]
+        col_ind_cos += [2*j-1, 2*j, 2*j-1, 2*j]
+
+        data_sin += [1, -1, -1, -1]
+        row_ind_sin += [2*j+2*nrot, 2*j+2*nrot-1, 2*nrot-2*j, 2*nrot-2*j-1]
+        col_ind_sin += [2*j-1, 2*j, 2*j-1, 2*j]
+
+    # Next do terms when j = n
+    data_cos += [1, 1, 1]
+    row_ind_cos += [4*nrot-1, 0, 4*nrot]
+    col_ind_cos += [2*nrot-1, 2*nrot, 2*nrot]
+
+    data_sin += [-1, 1, -1]
+    row_ind_sin += [0, 4*nrot, 4*nrot-1]
+    col_ind_sin += [2*nrot-1, 2*nrot-1, 2*nrot]
+
+    # Finish with terms when j > n
+    for j in range(nrot+1, num_freqs+1):
+
+        data_cos += [1, 1]
+        row_ind_cos += [2*j-2*nrot-1, 2*j-2*nrot]
+        col_ind_cos += [2*j-1, 2*j]
+
+        data_sin += [-1, 1]
+        row_ind_sin += [2*j-2*nrot, 2*j-2*nrot-1]
+        col_ind_sin += [2*j-1, 2*j]
+
+        if j + nrot <= num_freqs:
+
+            data_cos += [1, 1]
+            row_ind_cos += [2*j+2*nrot-1, 2*j+2*nrot]
+            col_ind_cos += [2*j-1, 2*j]
+
+            data_sin += [1, -1]
+            row_ind_sin += [2*j+2*nrot, 2*j+2*nrot-1]
+            col_ind_sin += [2*j-1, 2*j]
+
+    # Make sparse matrices
+    cos_matrix = sparse.csr_matrix(
+        (data_cos, (row_ind_cos, col_ind_cos)))
+    sin_matrix = sparse.csr_matrix(
+        (data_sin, (row_ind_sin, col_ind_sin)))
+
+    return cos_matrix, sin_matrix
```

Comment:
I wonder if this method can be made much faster. How large do you think these matrices will get? If they don't get past, say, 1000x1000, then it might be okay to just use dense matrices.

Reply:
Unfortunately these matrices can be quite large - the estimator accuracy is limited roughly to 2*pi/(matrix size), so I tend to run this up to ~10000 or even 100000. Luckily they're incredibly sparse (max 4 non-zero entries per row).